elastic/e2e-testing
# feat: support using local Beats repository in the tests

## What does this PR do?
It supports using the local copy of Beats repo to fetch the elastic-agent binaries that were already built from Beats build system.

It enables this feature with the `BEATS_LOCAL_PATH` environment variable. If the path defined in the variable exists will use it as base directory to build a path to the distributions dir, which is: `${BEATS_LOCAL_PATH}/x-pack/elastic-agent/build/distributions`.



## Why is it important?
It will allow developers not to wait for the packaging job to run the E2E tests locally, consuming local artifacts produced by their own build.




## Checklist




## Author's Checklist


- [x] Verified the build commands for the agent

## How to test this PR locally

1. build the agent from Beats
```shell
cd $BEATS_LOCAL_PATH
x-pack/elastic-agent; SNAPSHOT=true PLATFORMS="+all linux/amd64" DEV=true mage package
```

2. run the tests
```shell
SUITE=fleet TAGS="fleet_mode_agent && enroll" TIMEOUT_FACTOR=3 LOG_LEVEL=TRACE DEVELOPER_MODE=true BEATS_LOCAL_PATH=~/.op/git/beats make -C e2e functional-test
```

2a. run the tests from a non existen path
```shell
SUITE=fleet TAGS="fleet_mode_agent && enroll" TIMEOUT_FACTOR=3 LOG_LEVEL=TRACE DEVELOPER_MODE=true BEATS_LOCAL_PATH=~/.op/git/garbage make -C e2e functional-test
```
It should fail with:

>ERRO[2021-02-01T18:19:04+01:00] Could not download the binary for the agent   arch=x86_64 artifact=elastic-agent error="stat /Users/mdelapenya/.op/git/garbage/x-pack/elastic-agent/build/distributions/elastic-agent-8.0.0-SNAPSHOT-x86_64.rpm: no such file or directory" extension=rpm os=linux version=8.0.0-SNAPSHOT
FATA[2021-02-01T18:19:04+01:00] Sorry, we could not download the installer    error="stat /Users/mdelapenya/.op/git/garbage/x-pack/elastic-agent/build/distributions/elastic-agent-8.0.0-SNAPSHOT-x86_64.rpm: no such file or directory" image=centos installer=systemd
make: *** [functional-test] Error 1



## Related issues


- Closes #653 










## Follow-ups
At this moment it's not loading the Docker images for the stand-alone images. I need to figure out how to create a Docker installer that loads the image using the PreInstall functions in the installer. I wanted to share this code to receive feedback the soonest.
